lib/handler/cache-handler.js

```diff
@@ -0,0 +1,368 @@
+  }
+
+  /**
+   * @see {DispatchHandlers.onHeaders}
+   *
+   * @param {number} statusCode
+   * @param {Buffer[]} rawHeaders
+   * @param {() => void} resume
+   * @param {string} statusMessage
+   * @param {Record<string, string | string[]> | undefined} headers
+   * @returns {boolean}
+   */
+  onHeaders (
+    statusCode,
+    rawHeaders,
+    resume,
+    statusMessage,
+    headers = util.parseHeaders(rawHeaders)
+  ) {
+    if (
+      this.#req.method in UNSAFE_METHODS &&
+      statusCode >= 200 &&
+      statusCode <= 399
+    ) {
+      // https://www.rfc-editor.org/rfc/rfc9111.html#name-invalidating-stored-respons
+      const result = this.#opts.store.deleteByOrigin(this.#req.origin)
+      if (
+        result &&
+        typeof result.catch === 'function' &&
+        typeof this.#handler.onError === 'function'
+      ) {
+        result.catch(err => {
+          this.#handler.onError(err)
+        })
+      }
+
+      return this.#handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage,
+        headers
+      )
+    }
+
+    const cacheControlHeader = headers['cache-control']
+    const contentLengthHeader = headers['content-length']
+
+    if (!cacheControlHeader || !contentLengthHeader || this.#opts.store.isFull) {
+      // Don't have the headers we need, can't cache
+      return this.#handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage,
+        headers
+      )
+    }
+
+    const contentLength = Number(contentLengthHeader)
+    if (!Number.isInteger(contentLength)) {
+      return this.#handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage,
+        headers
+      )
+    }
+
+    const cacheControlDirectives = parseCacheControlHeader(cacheControlHeader)
+    if (!canCacheResponse(statusCode, headers, cacheControlDirectives)) {
+      return this.#handler.onHeaders(
+        statusCode,
+        rawHeaders,
+        resume,
+        statusMessage,
+        headers
+      )
+    }
+
+    const now = Date.now()
+    const staleAt = determineStaleAt(now, headers, cacheControlDirectives)
+    if (staleAt) {
+      const varyDirectives = headers.vary
+        ? parseVaryHeader(headers.vary, this.#req.headers)
+        : undefined
+      const deleteAt = determineDeleteAt(now, cacheControlDirectives, staleAt)
+
+      const strippedHeaders = stripNecessaryHeaders(
+        rawHeaders,
+        headers,
+        cacheControlDirectives
+      )
+
+      this.#writeStream = this.#opts.store.createWriteStream(this.#req, {
+        statusCode,
+        statusMessage,
+        rawHeaders: strippedHeaders,
+        vary: varyDirectives,
+        cachedAt: now,
+        staleAt,
+        deleteAt
+      })
+
+      if (this.#writeStream) {
+        this.#writeStream.on('drain', resume)
+        this.#writeStream.on('error', (err) => {
+          this.#handler.onError(err)
```

Comment:
```suggestion
```

There is not reason why this should fail the handler. Again some kind of warning or something?